File: algorithms/binary_search/binary_search.py
from random import randint

def binary_search(arr: list[int], ubound: int, value: int):
    lbound = 0

    while lbound >= 0 and lbound <= ubound:
        midpoint = int((lbound + ubound) / 2)
        if arr[midpoint] == value:
            return midpoint + 1
            
        if arr[midpoint] > value:
            ubound = midpoint - 1
        else:
            lbound = midpoint + 1

    return -1


def binary_search_recursive(
    arr: list[int], lbound: int, ubound: int, value: int,
):
    midpoint = int((lbound + ubound) / 2)
    if arr[midpoint] == value:
        return midpoint + 1
    if arr[midpoint] > value:
        return binary_search_recursive(arr, lbound, midpoint - 1, value)
    return binary_search_recursive(arr, midpoint + 1, ubound, value)


def main():
    arr = sorted([randint(0, 100) for _ in range(10)])
    value_to_find = randint(0, 100)

    found_index = binary_search(arr, len(arr) - 1, value_to_find)
    # uncomment the following line for the recursive version
    # binary_search_recursive(arr, 0, value_to_find)
    print(f"The array is: {arr}")
    print(f"The value to find is: {value_to_find}")
    print(f"Found value at position: {found_index}")


if __name__ == "__main__":
    main()

File: algorithms/binary_search/binary_search.c
#include<stdio.h>
#include<stdlib.h>


int binary_search(int *arr, int ubound, int val) {
    int lbound = 0;

    while (lbound >= 0 && lbound <= ubound) {
        int midpoint = (lbound + ubound) / 2;

        if (arr[midpoint] == val) {
            return midpoint + 1;
        }

        if (arr[midpoint] > val) {
            ubound = midpoint - 1;
        }
        else {
            lbound = midpoint + 1;
        }
    }
    return -1;
}


int binary_search_recursive(int *arr, int lbound, int ubound, int val) {
    int midpoint = (ubound + lbound) / 2;
    if (arr[midpoint] == val) {
        return midpoint + 1;
    }
    if (arr[midpoint] > val) {
        return binary_search_recursive(arr, lbound, midpoint - 1, val);
    }
    return binary_search_recursive(arr, midpoint + 1, ubound, val);
}


int main(int argc, char *argv[]) {
    int n = argc - 1;
    int *arr = malloc(n * sizeof(int));

    for (int i = 0; i < n; i++) {
        char *endPtr;
        arr[i] = strtol(argv[i + 1], &endPtr, 10);
    }

    int valToFind;
    printf("Enter an integer to find: ");
    scanf("%d", &valToFind);

    // int position = binary_search(arr, n, valToFind);
    int position = binary_search_recursive(
        arr, 0, n, valToFind
    );

    if (position == -1) {
        return 1;
    }

    printf("The position is: %i\n", position);

    free(arr);
    return 0;
}
